### PR TITLE
Issue131/gpl 560 as a developer id like to replace the view with a report rebase

### LINKED ISSUE #131: GPL-560 As a developer I would like to replace the materialised view with a MLWH report
Unified WH component of: https://github.com/sanger/samples_extraction/issues/118

**User story**
As a developer I would like to replace the materialised view  introduced in GPL-544 #117 with a report in the unified warehouse to reduce coupling to the samples_extraction database, increase flexibility of reporting, centralise the report tables, and reduce overall brittleness of the reports.

**Who are the primary contacts for this story**
James G (Wrote original view)
Laura L

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] Reporting on all data within the existing view
- [ ] Reporting of correct source barcodes for ALL activities

**Additional context**
The addition of the view was made to fulfil an urgent need to gain insight into Heron samples. However it has resulted in a fairly complicated query, which is difficult to generalise to all activities. The main difficult is in ensuring that the report can handle:
1) Activities in which there are intermediate step between the input plate and output
2) Activities where we are dealing with anything other than a single input mapped to a single output
3) Multiple different asset types (especially tubes vs. plate/racks)

The current view handles requirement 1, and has some support for requirement 3, however currently does not support requirement 2. To help restrict this impact it confines itself to ONLY the heron extraction pipleines, which at time of writing have other issues confining them to 1->1 transfers.

Meanwhile we cheat on Option 1, essentially extracting the source of the earliest transfer associated with the activity, this avoids the need for recursive queries, but prevents us simultaneously delivering requirement 2.

Meanwhile, a report in the application itself will be able to make use of internal application logic, and potentially handle recursive queries programatically.
The view also runs counter to the general strategy of building reports of the warehouse and isolating application databases within the applications themselves.

### PR DESCRIPTION
**Pull request by JamesGlover**
Closes #131 

Changes proposed in this pull request:

* Upgrade dependencies including:
  - Ruby -> 2.7.2
  - Rails -> 6.0.3
  - Rubocop 
* Strip down rails dependencies to actually used components
* Add sample extraction activity messages
* Run safe cops
* Clean up unused code

### CODE REVIEW

**Comment by JamesGlover** [comment]
Compare the key changes:
https://github.com/sanger/unified_warehouse/pull/133/files/53d3676f85932c763120c814b4558ef97d4d0bce..0ebc924279f8159141765d0d41567095e49dbd98
I've rebased to group Rails 6 associated changes and cops together. The above focuses on what should be the important changes. (Not saying don't look at the other changes, but these are the ones which aren't mainly automated)

File: README.md [hunk 1]
```diff
@@ -57,12 +56,17 @@ This will ensure the queue that is generated is compatible with the expectations
 
 ### Execution
 
-Execute the worker to pick up messages in the queue and process them into the database:
+Execute the worker to pick up messages in the queue and process them into the
+database:
 
-    bundle exec ./script/worker 1 start
+    ```bundle exec ./bin/amqp_client start```
```
Comment (JamesGlover) [inline README.md:63]: The script was a layer of abstraction added for the old deployment project, and isn't used in production any more. It just complicates things, and can hide useful error messages.

File: README.md [hunk 2]
```diff
@@ -12,19 +12,18 @@ This application is a denormalised warehouse for multiple LIMS.
...
 1. Install bundler the version of bundler used to create the `Gemfile.lock`:
 
-       gem install bundler -v $(tail -1 Gemfile.lock)
+       `gem install bundler -v $(tail -1 Gemfile.lock)`
 
-1. Install the relevant gems using bundler:
+1. Run the setup process:
 
-       bundle install
+       `bin/setup`
```
Comment (JamesGlover) [inline README.md:20]: I've been updating the bin/setup scripts as I use them, makes on-boarding developers a little easier.

File: README.md [hunk 3]
```diff
@@ -12,19 +12,18 @@ This application is a denormalised warehouse for multiple LIMS.
...
 
 ### Database preparation
 
-Before you can use the system in any capacity, you must first prepare the database:
+Before you can use the system in any capacity, you must first prepare the database.
+This should be handled by `bin/setup` above, but if not:
 
-1. `bundle exec rake db:create`
-1. `bundle exec rake db:structure:load`
-1. `bundle exec rake db:seed`
```
Comment (JamesGlover) [inline README.md:27]: Rails 6 handles structure:load automatically now.

File: app/models/samples_extraction_activity.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    ignore(
+      :samples
+    )
+
+    translate(
+      activity_id: :id_activity_lims
+    )
+  end
+end
```
Comment (JamesGlover) [inline app/models/samples_extraction_activity.rb:29]: And once all the yak shaving has been done, this, along with the migration, is the main change needed to actually deliver the story.

File: bin/amqp_client [hunk 5]
```diff
@@ -27,6 +27,7 @@ class AmqpClient
     end
 
     def start!
+      make_pid_dir
       worker_count.times do |i|
```
Comment (JamesGlover) [inline bin/amqp_client:31]: Spotted a troubleshooting section in the Readme that indicated the script would error if the pid directory didn't exist. So added a fix that will create it for us.

File: bin/amqp_client [hunk 6]
```diff
@@ -64,6 +65,14 @@ class AmqpClient
...
     end
+
+    private
+
+    def make_pid_dir
+      Dir.mkdir(@pid_dir) unless Dir.exist?(@pid_dir)
+    rescue Errno::EEXIST
+      # Race condition, another process has created the directory.
+    end
   end
```
Comment (JamesGlover) [inline bin/amqp_client:76]: Might happen if we're spinning up multiple workers at the same time.

File: bin/setup [hunk 7]
```diff
@@ -30,7 +30,4 @@ chdir APP_ROOT do
 
   puts "\n== Removing old logs and tempfiles =="
   system! 'bin/rails log:clear tmp:clear'
-
-  puts "\n== Restarting application server =="
-  system! 'bin/rails restart'
```
Comment (JamesGlover) [inline bin/setup]: Removed, as we don't have a server. Could of course restart our workers, but I've always found this bit of dubious merit anyway.

File: config/environments/development.rb [hunk 8]
```diff
@@ -74,7 +74,16 @@
...
+    # Test binding for development
+    'test.key',
+    # Sequencescape bindings
+    'development.saved.sample.#',
+    'development.saved.study.#',
+    'development.message.#',
+    # Sample Extraction bindings
+    'development.activity.finished.#'
+  ]
   config.amqp.main.deadletter_exchange    = 'deadletters'
```
Comment (JamesGlover) [inline config/environments/development.rb:87]: Again, inspired by comments in the Readme. I've added the keys we use for Sequencescape and sample extraction. No real reason not to streamline this part of the setup. (Especially in development mode.)

File: config/preinitializer.rb [hunk 9]
```diff
@@ -1,20 +0,0 @@
...
-end
-
-begin
-  # Set up load paths for all bundled gems
-  ENV["BUNDLE_GEMFILE"] = File.expand_path("../../Gemfile", __FILE__)
-  Bundler.setup
-rescue Bundler::GemNotFound
-  raise RuntimeError, "Bundler couldn't find some gems." +
-                      "Did you run `bundle install`?"
-end
```
Comment (JamesGlover) [inline config/preinitializer.rb:20]: Not needed any more. This is left-over from earlier warehouses and ancient Rails versions.

File: db/migrate/20190403081352_add_bmap_flowcell_table.rb [hunk 10]
```diff
@@ -2,7 +2,6 @@
 
 class AddBmapFlowcellTable < ActiveRecord::Migration
   def change
-    # rubocop:disable Rails/CreateTableWithTimestamps
     create_table :bmap_flowcell, primary_key: :id_bmap_flowcell_tmp do |t|
```
Comment (JamesGlover) [inline db/migrate/20190403081352_add_bmap_flowcell_table.rb:5]: We still disable these cops, but just do so globally. Avoids an unwitting developer getting confused by Rubocop insisting on something we don't actually want.

File: db/migrate/20201028091922_add_samples_extraction_activity_table.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+      t.datetime :last_updated, null: false, comment: 'Timestamp of last change to activity'
+      t.datetime :recorded_at, null: false, comment: 'Timestamp of warehouse update'
+      t.datetime :completed_at, null: false, comment: 'Timestamp of activity completion'
+      t.datetime :deleted_at, null: true, comment: 'Timestamp of any activity removal'
+
+      t.string :id_lims, limit: 10, null: false, comment: 'LIM system identifier'
+    end
+  end
+end
```
Comment (JamesGlover) [inline db/migrate/20201028091922_add_samples_extraction_activity_table.rb:27]: And the table itself.

File: db/structure.sql [hunk 12]
```diff
@@ -14,11 +9,17 @@
...
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!50503 SET character_set_client = utf8mb4 */;
+CREATE TABLE `ar_internal_metadata` (
+  `key` varchar(255) NOT NULL,
+  `value` varchar(255) DEFAULT NULL,
+  `created_at` datetime(6) NOT NULL,
+  `updated_at` datetime(6) NOT NULL,
+  PRIMARY KEY (`key`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+/*!40101 SET character_set_client = @saved_cs_client */;
```
Comment (JamesGlover) [inline db/structure.sql:22]: Rails 6 adds this for its own ends.

File: db/structure.sql [hunk 13]
```diff
@@ -28,15 +29,15 @@ CREATE TABLE `bmap_flowcell` (
   `recorded_at` datetime NOT NULL COMMENT 'Timestamp of warehouse update',
   `id_sample_tmp` int unsigned NOT NULL COMMENT 'Sample id, see "sample.id_sample_tmp"',
   `id_study_tmp` int unsigned NOT NULL COMMENT 'Study id, see "study.id_study_tmp"',
```
Comment (JamesGlover) [inline db/structure.sql:31]: Most of these changes are due to the frustratingly inconsistent ways in which different MySQL versions dump their table schemas.

Comment (JamesGlover) [reply db/structure.sql:31]: Ohh, should add, this won't affect the production warehouses at all. Just developer ones, and the CHARACTER SET option its dumped here matches what we have in production anyway.

File: lib/payload.rb [hunk 14]
```diff
@@ -36,7 +36,7 @@ def record
 
   def model_class_for(model_class_name)
     candidate_class = model_class_name.classify.safe_constantize
-    unless candidate_class && (candidate_class < ApplicationRecord)
+    unless candidate_class.respond_to?(:create_or_update_from_json)
       raise(InvalidMessage, "Unrecognized model: #{model_class_name}")
```
Comment (JamesGlover) [inline lib/payload.rb:40]: We currently have lots of deadletters from barcodes, which get routed to the WH, but are handled elsewhere. This lets the WH just ignore the messages.

File: lib/postman/channel.rb [hunk 15]
```diff
@@ -54,8 +54,10 @@ def queue
     end
 
     def queue_arguments
-      config = { 'x-dead-letter-exchange' => @deadletter_exchange }
+      config = {}
+      config['x-dead-letter-exchange'] = @deadletter_exchange
       config['x-message-ttl'] = @ttl if @ttl.present?
+      config['durable'] = true
       config
```
Comment (JamesGlover) [inline lib/postman/channel.rb:61]: More changes inspired by comments in the Readme. This should just reduce the number of tweaks needed to get things up and running.

File: lib/resource_tools/resource_table_migration.rb [hunk 16]
```diff
@@ -1,97 +0,0 @@
-module ResourceTools::ResourceTableMigration
```
Comment (JamesGlover) [inline lib/resource_tools/resource_table_migration.rb:1]: Handles migrations for warehouse_two, not used here.

File: Gemfile [hunk 17]
```diff
@@ -1,17 +1,37 @@
...
-gem "mysql2", "~> 0.4.10"
-gem "rails", "~> 4.1"
+gem 'bootsnap'
+gem 'mysql2'
 
-gem "bunny"
-gem "daemons"
-gem "hashie", "~> 3.3"
-gem "migration_comments"
+# We hardly use any of the features of Rails, so lets just load what we need
```
Comment (stevieing) [inline Gemfile:6]: Lots of commented out stuff. Can it just be removed?

File: config/initializers/application_controller_renderer.rb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
```
Comment (stevieing) [inline config/initializers/application_controller_renderer.rb:1]: Again. There are a few files in initializers with just comments. Can they be removed?

**Review by stevieing — APPROVED** [review]
All seems good. Just a couple of comments.

File: config/initializers/application_controller_renderer.rb (hunk 18 above)
Comment (JamesGlover) [reply config/initializers/application_controller_renderer.rb:1]: Again, yep, and actually did start stripping them out. Mostly left them there it make it easier to reinstate the behaviour later. But definitely happy to remove (especially as re-enabling isn't as simple as just uncommenting the code)

File: Gemfile (hunk 17 above)
Comment (JamesGlover) [reply Gemfile:6]: Yes, was mostly left there to direct future developers if they want to reinstate a feature, but may be better of just putting a comment with an example, and a pointer to the rails documentation.

**Pull request merged**